parser.js

```diff
@@ -124,56 +132,90 @@ function generateAST(code, options) {
     warnedAboutTSVersion = true;
   }
 
-  // Even if jsx option is set in typescript compiler, filename still has to
-  // contain .tsx file extension
-  const FILENAME = extra.ecmaFeatures.jsx ? 'estree.tsx' : 'estree.ts';
-
-  const compilerHost = {
-    fileExists() {
-      return true;
-    },
-    getCanonicalFileName() {
-      return FILENAME;
-    },
-    getCurrentDirectory() {
-      return '';
-    },
-    getDefaultLibFileName() {
-      return 'lib.d.ts';
-    },
-
-    // TODO: Support Windows CRLF
-    getNewLine() {
-      return '\n';
-    },
-    getSourceFile(filename) {
-      return ts.createSourceFile(filename, code, ts.ScriptTarget.Latest, true);
-    },
-    readFile() {
-      return null;
-    },
-    useCaseSensitiveFileNames() {
-      return true;
-    },
-    writeFile() {
-      return null;
-    }
-  };
+  let relevantProgram = undefined;
+  let ast = undefined;
+  const shouldProvideParserServices = extra.project && extra.project.length > 0;
 
-  const program = ts.createProgram(
-    [FILENAME],
-    {
-      noResolve: true,
-      target: ts.ScriptTarget.Latest,
-      jsx: extra.ecmaFeatures.jsx ? 'preserve' : undefined
-    },
-    compilerHost
-  );
+  if (shouldProvideParserServices) {
+    const FILENAME = options.filePath;
+    const programs = calculateProjectParserOptions(code, options, extra);
+    for (const program of programs) {
```

Comment:
Is it possible for multiple programs to own the same file? Do you have to instead filter the project list down and then operate on all projects at once? Disclaimer: I don't know what the right workflow is. 😄